### PR TITLE
User admin

### PR DESCRIPTION
**Pull request by Emiliocaamano**
## Summary

User admin role implementation

## Known Issues

None

## Trello Card

(https://trello.com/c/guiELwe9)

### CODE REVIEW

File: app/constants/errorMessages.js [hunk 1]
```diff
@@ -5,7 +5,9 @@ const errorMsg = {
   decritedError: 'Error: Password does not correspond',
   userFindError: 'Error: User is not found',
   emailError: 'Error: Email is not in the database',
-  tokenError: 'Error: Failed to create a token'
+  tokenError: 'Error: Failed to create a token',
+  signInError: 'Error: you need to sign in to be an admin',
+  adminError: 'Error: this user is already an admin'
```
Comment (javerBenavidez) [inline app/constants/errorMessages.js]: Where are you using those messages?

File: app/controllers/users.js [hunk 2]
```diff
@@ -41,3 +41,23 @@ exports.getUsers = async (req, res, next) => {
     next(error);
   }
 };
+exports.adminLog = async (req, res, next) => {
+  try {
+    const userObject = req.body;
+    const userFound = await userService.findUser(userObject);
+    if (userFound === null) {
```
Comment (javerBenavidez) [inline app/controllers/users.js:48]:
```suggestion
    if (!userFound) {
```

Read [this guide](https://github.com/Wolox/tech-guides/blob/master/nodejs/docs/node-standard-and-best-practices.md#41--ifs)

File: app/controllers/users.js [hunk 3]
```diff
@@ -41,3 +41,23 @@ exports.getUsers = async (req, res, next) => {
...
+      const password = helper.encryptPassword(req.body.password);
```
Comment (javerBenavidez) [inline app/controllers/users.js]: use `userObject` variable instead of `req.body`

Comment (javerBenavidez) [inline app/controllers/users.js]: `passEncryptred` could be better

File: app/controllers/users.js [hunk 4]
```diff
@@ -41,3 +41,23 @@ exports.getUsers = async (req, res, next) => {
...
+      const userCreated = varParse.nameParse(userObject, password);
+      await userService.createUser(userCreated);
+      const admin = await helper.adminRole(userCreated);
```
Comment (javerBenavidez) [inline app/controllers/users.js]: You are saving the user before you change the admin property. 
You should change the property admin to true and then save de user

Comment (javerBenavidez) [inline app/controllers/users.js]: And you can add the `admin` property in `nameParse` method

File: app/controllers/users.js [hunk 5]
```diff
@@ -41,3 +41,23 @@ exports.getUsers = async (req, res, next) => {
...
+      logger.info('Admin created: ', admin);
+      res.status(201).send(admin);
```
Comment (javerBenavidez) [inline app/controllers/users.js]: remember not to log the user password in console. Remove password property from the user object

File: app/controllers/users.js [hunk 6]
```diff
@@ -41,3 +41,23 @@ exports.getUsers = async (req, res, next) => {
...
+    } else {
+      const admin = await helper.adminRole(userFound);
+      res.status(200).send(admin);
```
Comment (javerBenavidez) [inline app/controllers/users.js]: if the user is found you are not updating the object in db

File: app/helpers/helpers.js [hunk 7]
```diff
@@ -34,3 +34,8 @@ exports.createToken = userOnline => {
     throw defaultError(errorMsg.tokenError);
   }
 };
+exports.adminRole = usuario => {
+  usuario.admin = true;
+  console.log(usuario);
```
Comment (javerBenavidez) [inline app/helpers/helpers.js]: delete console.log

File: app/helpers/helpers.js [hunk 8]
```diff
@@ -34,3 +34,8 @@ exports.createToken = userOnline => {
...
+  return usuario;
+};
```
Comment (javerBenavidez) [inline app/helpers/helpers.js]: you can do this in the serializer file

File: app/models/users.js [hunk 9]
```diff
@@ -58,7 +58,8 @@ module.exports = (sequelize, DataTypes) => {
             msg: 'La contraseña tiene que tener minimamente 8 caracteres'
           }
         }
-      }
+      },
+      admin: false
```
Comment (javerBenavidez) [inline app/models/users.js]: What type of data is admin? It is allowNull? What value could be by default?

File: app/services/users.js [hunk 10]
```diff
@@ -34,3 +34,19 @@ exports.getUsers = async () => {
...
+//         firstName,
+//         lastName
+//       }
+//     });
+//     return userFound;
+//   } catch (error) {
+//     logger.error(error);
+//     throw databaseError(errorMsg.userFindError);
+//   }
+// };
```
Comment (javerBenavidez) [inline app/services/users.js]: If this is not necessary remove commented code

File: migrations/migrations/20220906155837-user-table.js [hunk 11]
```diff
@@ -26,6 +26,9 @@ module.exports = {
       password: {
         type: Sequelize.STRING,
         allowNull: false
+      },
+      admin: {
+        type: Sequelize.BOOLEAN
```
Comment (javerBenavidez) [inline migrations/migrations/20220906155837-user-table.js]: What type of data is admin? It is allowNull? What value could be by default?

File: test/__test__/signUp.test.js [hunk 12]
```diff
@@ -3,7 +3,7 @@ const app = require('../../app');
 const userService = require('../../app/services/users');
 const mockSignUp = require('../mockSignUp');
 
-describe('Should get a user', () => {
+describe.skip('Should get a user', () => {
   let userServiceMock = 0;
```
Comment (javerBenavidez) [inline test/__test__/signUp.test.js]: Take care when you put this `.skip()`. When you are going to push code, delete it.

File: test/__test__/userAdmin.test.js [hunk 13]
```diff
@@ -0,0 +1,28 @@
+const request = require('supertest');
+const app = require('../../app');
+const User = require('../../app/models').users;
+const mockUserAdmin = require('../mockUserAdmin');
+const helper = require('../../app/helpers/helpers');
+
+describe('Should get a token', () => {
```
Comment (javerBenavidez) [inline test/__test__/userAdmin.test.js]: Change the description, the same with the other tests.
Should have a general description of file you are testing.

File: test/__test__/userAdmin.test.js [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+  let findOneMock = 0;
+  let adminRoleMock = 0;
+  beforeEach(() => {
+    adminRoleMock = jest.spyOn(helper, 'adminRole').mockResolvedValueOnce(mockUserAdmin.mock1);
+    findOneMock = jest.spyOn(User, 'findOne').mockRejectedValueOnce(mockUserAdmin.mock2);
+  });
+  test('Should log all user', async () => {
```
Comment (javerBenavidez) [inline test/__test__/userAdmin.test.js]: take care with descriptions of each test

File: test/__test__/userAdmin.test.js [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+const mockUserAdmin = require('../mockUserAdmin');
+const helper = require('../../app/helpers/helpers');
+
+describe('Should get a token', () => {
+  let findOneMock = 0;
+  let adminRoleMock = 0;
+  beforeEach(() => {
+    adminRoleMock = jest.spyOn(helper, 'adminRole').mockResolvedValueOnce(mockUserAdmin.mock1);
+    findOneMock = jest.spyOn(User, 'findOne').mockRejectedValueOnce(mockUserAdmin.mock2);
+  });
```
Comment (javerBenavidez) [inline test/__test__/userAdmin.test.js]: When you are mocking some functions, and you have just one or two tests `beforeEach` is not necessary, you could put those mocks in each test

**Review by javerBenavidez — CHANGES_REQUESTED** [review]

File: app/controllers/users.js [hunk 16]
```diff
@@ -1,13 +1,14 @@
 const userService = require('../services/users');
-const varParse = require('../serializers/varParse');
+const varParse = require('../serializers/parser');
 const logger = require('../logger');
 const helper = require('../helpers/helpers');
+const parser = require('../serializers/parser');
```
Comment (javerBenavidez) [inline app/controllers/users.js]: varParse and parser is the same?

File: app/controllers/users.js [hunk 17]
```diff
@@ -1,13 +1,14 @@
...
 
 exports.signUp = async (req, res, next) => {
   try {
-    const password = helper.encryptPassword(req.body.password);
+    const password = helper.passEncryptred(req.body.password);
     const userObject = req.body;
```
Comment (javerBenavidez) [inline app/controllers/users.js]:
```suggestion
    const userObject = req.body;
    const password = helper.passEncryptred(userObject.password);
```

File: app/controllers/users.js [hunk 18]
```diff
@@ -22,7 +23,7 @@ exports.signIn = async (req, res, next) => {
     logger.info('User Log: ', userLog);
     const userLoged = await userService.findUser(userLog);
     logger.info('User Loged: ', userLoged);
-    const decryptedUser = await helper.decryptPassword(userLog.password, userLoged);
+    const decryptedUser = await helper.passEncryptred(userLog.password, userLoged);
```
Comment (javerBenavidez) [inline app/controllers/users.js]: why `passEncrypted` in `signIn` method?

File: app/controllers/users.js [hunk 19]
```diff
@@ -41,3 +42,24 @@ exports.getUsers = async (req, res, next) => {
     next(error);
   }
 };
+exports.adminLog = async (req, res, next) => {
+  try {
+    const userObject = req.body;
+    const userFound = await userService.findUser(userObject);
+    if (userFound === null) {
```
Comment (javerBenavidez) [inline app/controllers/users.js:48]:
```suggestion
    if (!userFound) {
```

File: app/controllers/users.js [hunk 20]
```diff
@@ -41,3 +42,24 @@ exports.getUsers = async (req, res, next) => {
...
+      const password = helper.passEncryptred(userObject.password);
+      const userCreated = varParse.nameParse(userObject, password);
+      const admin = await helper.adminRole(userCreated);
+      await userService.createUser(admin);
+      logger.info('Admin created: ', admin.firstName);
+      res.status(201).send(admin);
```
Comment (javerBenavidez) [inline app/controllers/users.js]: En el método nameParse() pusimos la variable `admin` como argumento, para que cuando le enviemos ese argumento en `true`, ya organice el código para ser almacenado.. Habiendo dicho esto, entonces lo podemos dejar como: 

```suggestion
      const userToCreate = varParse.nameParse(userObject, password, true);
      await userService.createUser(userToCreate);
      logger.info('Admin created: ', userToCreate.firstName);
      res.status(201).send(userToCreate);
```

File: app/controllers/users.js [hunk 21]
```diff
@@ -41,3 +42,24 @@ exports.getUsers = async (req, res, next) => {
...
+    } else {
+      const admin = await parser.adminRole(userFound);
```
Comment (javerBenavidez) [inline app/controllers/users.js]: utiliza la función que ya existe:

```suggestion
      const admin = await varParse.nameParse(userFound, userFound.password, true);
```

File: app/serializers/parser.js [hunk 22]
```diff
@@ -0,0 +1,11 @@
...
+  first_name: userParams.firstName,
+  last_name: userParams.lastName,
+  email: userParams.email,
+  password,
+  admin
+});
+exports.adminRole = usuario => {
+  usuario.admin = true;
+  return usuario;
+};
```
Comment (javerBenavidez) [inline app/serializers/parser.js]: remove this method

**Review by javerBenavidez — CHANGES_REQUESTED** [review]

File: app/controllers/users.js (hunk 19 above)
Comment (Emiliocaamano) [reply app/controllers/users.js:48]: There is a problem with that negated if, lint doesn't allow negeted conditions

Comment (javerBenavidez) [reply app/controllers/users.js:48]: Try putting whit out parenthesis
```suggestion
    if !userFound {
```

Comment (Emiliocaamano) [reply app/controllers/users.js:48]: it expect a parenthesis

Comment (javerBenavidez) [reply app/controllers/users.js:48]: ok ok, not problem with this comment

File: test/__test__/signUp.test.js (hunk 12 above)
Comment (javerBenavidez) [reply test/__test__/signUp.test.js]: @Emiliocaamano Te falta resolver este comentario

Comment (Emiliocaamano) [reply test/__test__/signUp.test.js]: Done!

**Review by javerBenavidez — APPROVED** [review]

**Pull request merged**